src/realm_dart_sync.cpp

```diff
@@ -138,3 +137,57 @@ RLM_API void realm_dart_sync_on_subscription_state_changed_callback(realm_userda
         (reinterpret_cast<realm_sync_on_subscription_state_changed_t>(ud->dart_callback))(ud->handle, state);
     });
 }
+
+bool wait_dart_before_continue(realm::util::UniqueFunction<void(realm::util::UniqueFunction<void(bool)>*)>* userCallback)
+{
+    std::condition_variable cv;
+    std::mutex m;
+    bool complete = false;
+    bool success = false;
+
+    realm::util::UniqueFunction<void(bool)> unlockFunc = [&complete, &success, &cv, &m](bool result) {
+        std::unique_lock<std::mutex> lk(m);
+        complete = true;
+        success = result;
+        lk.unlock();
+        cv.notify_one();
+    };
+    (*userCallback)(&unlockFunc);
+
+    std::unique_lock<std::mutex> lk(m);
+    cv.wait(lk, [&complete] {return complete; });
+
+    return success;
+}
+
+RLM_API bool realm_dart_sync_before_reset_handler_callback(realm_userdata_t userdata, realm_t* realm)
+{
+    auto ud = reinterpret_cast<realm_dart_userdata_async_t>(userdata);
+    realm::util::UniqueFunction<void(realm::util::UniqueFunction<void(bool)>*)> userCallback =
+        [&ud, &realm](realm::util::UniqueFunction<void(bool)>* unlockFunc) {
+        ud->scheduler->invoke([ud, &realm, unlockFunc]() {
+            (reinterpret_cast<realm_sync_before_client_reset_begin_func_t>(ud->dart_callback))(ud->handle, realm, unlockFunc);
+        });
+    };
```

Comment:
```suggestion
    realm::util::UniqueFunction userCallback =
        [ud, realm](realm::util::UniqueFunction<void(bool)>* unlockFunc) {
        ud->scheduler->invoke([ud, realm, unlockFunc]() {
            (reinterpret_cast<realm_sync_before_client_reset_begin_func_t>(ud->dart_callback))(ud->handle, realm, unlockFunc);
        });
    };
```
Don't pass argument by reference to escaping lambdas.. and let the compiler to the type inference.